server/dataset/cxg_dataset.py

```diff
@@ -50,12 +55,431 @@ def __init__(self, data_locator, app_config=None):
+                ordered = bool(json.loads(ordered_meta))
+            except (TypeError, json.JSONDecodeError):
+                ordered = False
+
+        categorical = pd.Categorical.from_codes(raw, categories=categories, ordered=ordered)
+        return pd.Series(categorical, name=column_name)
+
+    def _load_obs_annotations_from_storage(self, user_id: str) -> Optional[pd.DataFrame]:
+        user_root_uri = self._user_root_uri(user_id)
+        arrays = self._list_annotation_arrays(user_root_uri)
+        if not arrays:
+            return None
+
+        columns = {}
+        for column_name, array_uri in arrays.items():
+            series = self._load_annotation_column(array_uri)
+            if series is not None:
+                columns[column_name] = series.reset_index(drop=True)
+
+        if not columns:
+            return None
+
+        return pd.DataFrame(columns)
+
+    def _genesets_uri(self, user_id: str) -> str:
+        return path_join(self._user_root_uri(user_id), "genesets.json")
+
+    def _write_user_genesets(self, user_id: str, payload: Dict[str, Any]) -> None:
+        user_root_uri = self._ensure_user_storage(user_id)
+        genesets_uri = path_join(user_root_uri, "genesets.json")
+        vfs = tiledb.VFS(ctx=self.tiledb_ctx)
+        serialized = json.dumps(payload)
+        with vfs.open(genesets_uri, "wb") as handle:
+            handle.write(serialized.encode("utf-8"))
+
+    def _read_user_genesets(self, user_id: str) -> Optional[Dict[str, Any]]:
+        genesets_uri = self._genesets_uri(user_id)
+        vfs = tiledb.VFS(ctx=self.tiledb_ctx)
+        if not vfs.is_file(genesets_uri):
+            return None
+        try:
+            with vfs.open(genesets_uri, "rb") as handle:
+                raw = handle.read()
+        except TileDBError:
+            return None
+
+        try:
+            return json.loads(raw.decode("utf-8"))
+        except (UnicodeDecodeError, json.JSONDecodeError):
+            logging.warning("Failed to parse genesets payload at %s", genesets_uri)
+            return None
+
     def cleanup(self):
         """close all the open tiledb arrays"""
         for array in self.arrays.values():
             array.close()
         self.arrays.clear()
 
+    def save_obs_annotations(
+        self,
+        dataframe: pd.DataFrame,
+        user_id: Optional[str] = None,
+    ) -> None:
+        if user_id is None:
+            raise ValueError("User ID is required for saving annotations")
+
+        n_obs, _ = self.get_shape()
+        if n_obs <= 0:
+            return
+
+        user_root_uri = self._ensure_user_storage(user_id)
+        existing = self._list_annotation_arrays(user_root_uri)
+
+        if dataframe is not None and not dataframe.empty:
+            for column in dataframe.columns:
+                self._write_annotation_column(user_root_uri, column, dataframe[column], n_obs)
+                existing.pop(column, None)
+
+    def delete_obs_annotation_category(
```

Comment:
Persistence path: atomic write/move and copy‑delete as fallback. Please call out failure recovery guarantees; confirm no partial state is observable to readers.